### PR TITLE
ticdc: support graceful upgrade TiCDC pods

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve?

This PR supports upgrade TiCDC pods without causing latency spike.

Close https://github.com/pingcap/tidb-operator/issues/4646

### What is changed and how does it work?

Before update TiCDC statefuleset, we first drain the TiCDC in that pod. 

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [x] Manual test 
- [ ] No code 

| Upgrade TiCDC via editting config  |
| - |
| image |

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [x] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Support graceful upgrade TiCDC
```

### CODE REVIEW

File: pkg/manager/member/ticdc_upgrader.go [hunk 1]
```diff
@@ -108,7 +108,14 @@ func (u *ticdcUpgrader) Upgrade(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulS
 			}
 			continue
 		}
-		mngerutils.SetUpgradePartition(newSet, i)
+
+		err = gracefulShutdownTiCDC(tc, u.deps.CDCControl, u.deps.PodControl, pod, ordinal, "Upgrade")
+		if err != nil {
+			return err
+		}
```
Comment (KanShiori) [inline pkg/manager/member/ticdc_upgrader.go:115]: If only one ticdc, will wait until the timeout?

File: pkg/manager/member/ticdc_upgrader.go [hunk 2]
```diff
@@ -108,7 +108,14 @@ func (u *ticdcUpgrader) Upgrade(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulS
 			}
 			continue
 		}
-		mngerutils.SetUpgradePartition(newSet, i)
+
+		err = gracefulShutdownTiCDC(tc, u.deps.CDCControl, u.deps.PodControl, pod, ordinal, "Upgrade")
+		if err != nil {
+			return err
```
Comment (KanShiori) [inline pkg/manager/member/ticdc_upgrader.go:114]: add some description about the error

**Review by KanShiori — APPROVED** [review]

File: pkg/manager/member/ticdc_upgrader.go (hunk 1 above)
Comment (overvenus) [reply pkg/manager/member/ticdc_upgrader.go:115]: No, it skips graceful shutdown, see

https://github.com/pingcap/tidb-operator/blob/6c3bfa34dfdf7df4bfeb2502928c9b1df75d273c/pkg/controller/ticdc_control.go#L53-L65

File: pkg/manager/member/ticdc_upgrader.go (hunk 2 above)
Comment (overvenus) [reply pkg/manager/member/ticdc_upgrader.go:114]: We log details in `gracefulShutdownTiCDC`, see sample logs below.

```log
I0801 08:49:25.977589       1 event.go:282] Event(v1.ObjectReference{Kind:"TidbCluster", Namespace:"default", Name:"basic", UID:"16034049-84c9-4749-bde4-3008e10b59f5", APIVersion:"pingcap.com/v1alpha1", ResourceVersion:"2458961", FieldPath:""}): type: 'Normal' reason: 'Successfully Create' create ConfigMap/basic-ticdc-3436353 for controller TidbCluster/basic successfully
I0801 08:49:25.987193       1 stateful_set_control.go:94] : [default/basic]'s StatefulSet: [default/basic-ticdc] updated successfully
I0801 08:49:26.011110       1 tidbcluster_control.go:69] TidbCluster: [default/basic] updated successfully
I0801 08:49:26.949297       1 statefulset.go:169] set default/basic-ticdc partition to 3
I0801 08:49:26.949343       1 ticdc_scaler.go:195] ticdc.Upgrade: begin graceful shutdown basic-ticdc-2 in cluster default/basic
I0801 08:49:26.966106       1 pod_control.go:87] Pod: [default/basic-ticdc-2] updated successfully, : [default/basic]
I0801 08:49:26.966137       1 ticdc_scaler.go:207] ticdc.Upgrade: set pod basic-ticdc-2 in cluster default/basic annotation tidb.pingcap.com/ticdc-graceful-shutdown-begin-time to 2022-08-01T08:49:26Z successfully
I0801 08:49:27.124554       1 tidbcluster_control.go:69] TidbCluster: [default/basic] updated successfully
I0801 08:49:27.125517       1 tidb_cluster_controller.go:124] TidbCluster: default/basic, still need sync: ticdc.Upgrade: cluster default/basic basic-ticdc-2 is still the owner, try resign owner again, requeuing
I0801 08:49:28.548035       1 statefulset.go:169] set default/basic-ticdc partition to 3
I0801 08:49:29.892315       1 tidbcluster_control.go:69] TidbCluster: [default/basic] updated successfully
I0801 08:49:29.892351       1 tidb_cluster_controller.go:124] TidbCluster: default/basic, still need sync: ticdc.Upgrade: cluster default/basic basic-ticdc-2 still has 131 tables, wait draining, requeuing
I0801 08:49:30.147198       1 statefulset.go:169] set default/basic-ticdc partition to 3
I0801 08:49:31.534393       1 tidbcluster_control.go:69] TidbCluster: [default/basic] updated successfully
I0801 08:49:31.534444       1 tidb_cluster_controller.go:124] TidbCluster: default/basic, still need sync: ticdc.Upgrade: cluster default/basic basic-ticdc-2 still has 131 tables, wait draining, requeuing
I0801 08:49:31.747176       1 statefulset.go:169] set default/basic-ticdc partition to 3
I0801 08:49:32.873400       1 tidb_cluster_controller.go:124] TidbCluster: default/basic, still need sync: ticdc.Upgrade: cluster default/basic basic-ticdc-2 still has 131 tables, wait draining, requeuing
I0801 08:49:33.547769       1 statefulset.go:169] set default/basic-ticdc partition to 3
I0801 08:49:34.871809       1 tidb_cluster_controller.go:124] TidbCluster: default/basic, still need sync: ticdc.Upgrade: cluster default/basic basic-ticdc-2 still has 131 tables, wait draining, requeuing
I0801 08:49:36.547747       1 statefulset.go:169] set default/basic-ticdc partition to 3
I0801 08:49:37.872539       1 tidb_cluster_controller.go:124] TidbCluster: default/basic, still need sync: ticdc.Upgrade: cluster default/basic basic-ticdc-2 still has 131 tables, wait draining, requeuing
I0801 08:49:40.951512       1 statefulset.go:169] set default/basic-ticdc partition to 3
I0801 08:49:42.072695       1 tidb_cluster_controller.go:124] TidbCluster: default/basic, still need sync: ticdc.Upgrade: cluster default/basic basic-ticdc-2 still has 121 tables, wait draining, requeuing
I0801 08:50:06.049524       1 statefulset.go:169] set default/basic-ticdc partition to 3
I0801 08:50:07.272154       1 tidb_cluster_controller.go:124] TidbCluster: default/basic, still need sync: ticdc.Upgrade: cluster default/basic basic-ticdc-2 still has 101 tables, wait draining, requeuing
I0801 08:50:36.055293       1 statefulset.go:169] set default/basic-ticdc partition to 3
I0801 08:50:37.072982       1 tidb_cluster_controller.go:124] TidbCluster: default/basic, still need sync: ticdc.Upgrade: cluster default/basic basic-ticdc-2 still has 81 tables, wait draining, requeuing
I0801 08:50:46.159003       1 statefulset.go:169] set default/basic-ticdc partition to 3
I0801 08:50:47.272882       1 tidb_cluster_controller.go:124] TidbCluster: default/basic, still need sync: ticdc.Upgrade: cluster default/basic basic-ticdc-2 still has 71 tables, wait draining, requeuing
I0801 08:51:06.070187       1 statefulset.go:169] set default/basic-ticdc partition to 3
I0801 08:51:07.273668       1 tidb_cluster_controller.go:124] TidbCluster: default/basic, still need sync: ticdc.Upgrade: cluster default/basic basic-ticdc-2 still has 61 tables, wait draining, requeuing
I0801 08:51:36.052212       1 statefulset.go:169] set default/basic-ticdc partition to 3
I0801 08:51:37.076171       1 tidb_cluster_controller.go:124] TidbCluster: default/basic, still need sync: ticdc.Upgrade: cluster default/basic basic-ticdc-2 still has 31 tables, wait draining, requeuing
```

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 3]
```diff
@@ -680,6 +680,12 @@ type TiCDCSpec struct {
 	// Defaults to Kubernetes default storage class.
 	// +optional
 	StorageClassName *string `json:"storageClassName,omitempty"`
+
+	// GracefulShutdownTimeout is the timeout of gracefully shutdown a TiCDC pod.
+	// Encoded in the format of Go Duration.
+	// Defaults to 10m
+	// +optional
+	GracefulShutdownTimeout *string `json:"gracefulShutdownTimeout,omitempty"`
```
Comment (liubog2008) [inline pkg/apis/pingcap/v1alpha1/types.go:688]: How about using `metav1.Duration`?

**Review by liubog2008 — APPROVED** [review]

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**